### ISSUE TITLE
[CSL-314] Use snapshot when multiple reads must be done for the same version

### ISSUE THREAD

**Issue by jagajaga** [input-output-hk/cardano-sl#685]
<blockquote>@gromak</blockquote>

For example, in `Pos.Modern.Txp.Storage.Storage.processTx`.
It's hard to estimate this task, because we have never used snapshots and some difficulties can arise. If everything goes smoothly, it should take less than 2 hours of course.

**Comment by jagajaga** [comment]
<blockquote>@gromak</blockquote>

Another place where it probably makes sense to use snapshot is `normalizeTxpLD`. However, if we do it under lock on block application, then it's not necessary. Currently we don't have this lock at all.

**Comment by jagajaga** [comment]
<blockquote>@georgeee</blockquote>

Gromak: almost impossible to reproduce, though possible

**Comment by jagajaga** [comment]
<blockquote>@gromak</blockquote>

Snapshot is necessary in two cases:
1. In block verification.
2. In `Data` message verification.

**Comment by jagajaga** [comment]
<blockquote>@gromak</blockquote>

`processTx` is an example of (2).

**Comment by jagajaga** [comment]
<blockquote>@gromak</blockquote>

1. In block verification.
Currently it's not needed, because verification useless lock. It's bad and there is issue about it: CSL-600.
2. In Data message verification.
Usually it's not necessary.
For example, it's not necessary in `Delegation` and `GodTossing`, because there verification doesn't use DB at all.
Also it's not necessary in `Update` and `Txp`, but for another reason. These components store tip in memory state. Also `Update` currently uses lock, maybe it will be improved, maybe not, but it doesn't really matter.
Suppose we receive data which needs to be verified. In `Txp` we do all queries to DB first and then do verification and modifications in `STM`. Suppose that after we started querying DB and before we finished, tip has changed. It means that that we may consider invalid data valid. But in reality nothing bad can happen. Even if we put this data into memory state, we know that we will normalize it before releasing lock. And we can't create block without lock. So having bad data in mempool for a short interval won't have any bad consequences.
————————————————————————————————
Snapshots have been implemented long ago. After that it appeared that we don't need them in block verification/data processing. They will be needed to optimize LRC and for CSL-600, but these are separate issues.
